src/databricks/labs/ucx/migration/azure_credentials.py

```diff
@@ -0,0 +1,170 @@
+
+        for azure_sp_info in azure_sp_infos:
+            if azure_sp_info.secret_scope is None:
+                continue
+            if azure_sp_info.secret_key is None:
+                continue
+
+            try:
+                secret_response = self._ws.secrets.get_secret(azure_sp_info.secret_scope, azure_sp_info.secret_key)
+            except Unauthenticated:
+                logger.info(f"User is unauthenticated to fetch secret value. Cannot fetch the service principal "
+                            f"client_secret for {azure_sp_info.application_id}. Will not reuse this client_secret")
+                continue
+            except PermissionDenied:
+                logger.info(f"User does not have permission to read secret value for {azure_sp_info.secret_scope}.{azure_sp_info.secret_key}. "
+                            f"Cannot fetch the service principal client_secret for {azure_sp_info.application_id}. "
+                            f"Will not reuse this client_secret")
+                continue
+            except ResourceDoesNotExist:
+                logger.info(f"Secret {azure_sp_info.secret_scope}.{azure_sp_info.secret_key} does not exists. "
+                            f"Cannot fetch the service principal client_secret for {azure_sp_info.application_id}. "
+                            f"Will not reuse this client_secret")
+                continue
+            except InternalError:
+                logger.info(f"InternalError while reading secret {azure_sp_info.secret_scope}.{azure_sp_info.secret_key}. "
+                            f"Cannot fetch the service principal client_secret for {azure_sp_info.application_id}. "
+                            f"Will not reuse this client_secret")
+                continue
+
+            # decode the bytes string from GetSecretResponse to utf-8 string
+            # TODO: handle different encoding if we have feedback from the customer
+            try:
+                secret_value = base64.b64decode(secret_response.value).decode("utf-8")
+                azure_sp_info_with_client_secret.update(azure_sp_info.application_id, secret_value)
+            except UnicodeDecodeError:
+                logger.info(f"Secret {azure_sp_info.secret_scope}.{azure_sp_info.secret_key} has Base64 bytes that cannot be decoded to utf-8 string . "
+                            f"Cannot fetch the service principal client_secret for {azure_sp_info.application_id}. "
+                            f"Will not reuse this client_secret")
+
+        # update the list of ServicePrincipalMigrationInfo if client_secret is found
+        for sp in sp_list:
+            if sp.client_id in azure_sp_info_with_client_secret:
+                yield ServicePrincipalMigrationInfo.from_storage_permission_mapping(sp, azure_sp_info_with_client_secret[sp.client_id])
+
+
+    def _save_action_plan(self, sp_list_with_secret) -> str | None:
+        # save action plan to a file for customer to review.
```

Comment:
We may need that file, but for the workflow to pickup.

Reply:
so we may want the workflow to execute the migration, not just using cli? 
For now, I just removed the save action plan code, just print the info to console for user to confirm.